crates/amaru-kernel/src/memoized.rs

```diff
@@ -0,0 +1,265 @@
+use pallas_addresses::Address;
+use pallas_codec::{minicbor as cbor, minicbor::data::IanaTag, utils::Bytes};
+use pallas_crypto::hash::{Hash, Hasher};
+use pallas_primitives::{
+    alonzo,
+    conway::{
+        KeepRaw, MintedDatumOption, MintedTransactionOutput, NativeScript, PseudoScript, Value,
+    },
+    DatumHash, NonEmptyKeyValuePairs, PlutusData,
+};
+
+// --------------------------------------------------- MemoizedTransactionOutput
+
+#[derive(Debug, Clone)]
+pub struct MemoizedTransactionOutput {
+    is_legacy: bool,
+    pub address: Address,
+    pub value: Value,
+    pub datum: MemoizedDatum,
+    pub script: Option<MemoizedScript>,
+}
+
+impl<'b, C> cbor::Decode<'b, C> for MemoizedTransactionOutput {
+    // TODO: This implementation still copies the bytes when we convert from the minted version to
+    // the owned version. To avoid that, we should actually rewrite the decoders for all the
+    // memoized structures, so that they can retain their original bytes.
+    fn decode(d: &mut cbor::Decoder<'b>, ctx: &mut C) -> Result<Self, cbor::decode::Error> {
+        let minted: MintedTransactionOutput<'_> = d.decode_with(ctx)?;
+        Self::try_from(minted).map_err(cbor::decode::Error::message)
+    }
+}
+
+impl<C> cbor::Encode<C> for MemoizedTransactionOutput {
+    fn encode<W: cbor::encode::Write>(
```

Comment:
I understand what's going on here since I've spent so much time staring at the CDDL recently, but it is probably helpful to provide some comments here providing context to the next reader (or ourselves in 6 months at 3am trying to fix a bug...).

Maybe even just a "the cbor encoding for a transaction output is defined [here](https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/conway/impl/cddl-files/conway.cddl#L162)"